### PR TITLE
🪟 🎉 Free connector program enrollment modal

### LINKED ISSUE #21224: [Free connectors] create enrollment modal
Part of https://github.com/airbytehq/airbyte/issues/21074. Clicking on the enrollment CTA banners (i.e. https://github.com/airbytehq/airbyte/issues/21222 or https://github.com/airbytehq/airbyte/issues/21223) triggers a modal. The modal's content explains the program and enrollment requirements (i.e. registering future-billable payment info) in more detail, and it contains the CTA button which actually sends users to the stripe connect setup flow. For airbyte employees, a Figma mockup can be found [here](https://www.figma.com/file/wihAJg5mkp61I7lLHs0IpV/FREE-CONNECTOR-PROGRAM?node-id=1106%3A23359&t=sv3jVkm9Yn7QlwMU-0); it looks like this
2023-01-10-11:32:12-screenshot

Users enter the stripe connect setup flow via an internal url expects to be given a `successUrl` and `cancelUrl` parameter and will set up the slack session and redirect the user. For now, the success and cancel urls should be the user's current url with distinct query params appended (e.g. `?freeConnectorProgramEnrollment=success` or `?freeConnectorProgramEnrollment=cancel`).

### PR DESCRIPTION
**Pull request by joeykmh**
## What
Closes https://github.com/airbytehq/airbyte/issues/21224

Adds components for the enrollment modal for the free connector program:

image

Figma designs: https://www.figma.com/file/wihAJg5mkp61I7lLHs0IpV/FREE-CONNECTOR-PROGRAM?node-id=1106%3A23359&t=sv3jVkm9Yn7QlwMU-0

## How
- Adds `FreeConnectorProgramEnrollmentButton`, which contains all the logic required to render the modal.
- Adjusts the `createCheckout()` method in our `StripeService` to accept the new "setup" mode
- Does not yet display this button anywhere

The idea is that this button can be used in various other components where an "Enroll now!" CTA is required.

## Recommended reading order
Top to bottom

## Considerations
`FreeConnectorProgramEnrollmentButton` is a very long name 😄  Maybe we should use a `FCP*` prefix for these components? Long names don't bother me too much, so I don't feel super strongly.

### CODE REVIEW

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/EnrollmentModal.tsx
Comment (josephkmh) [inline airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/EnrollmentModal.tsx:46]: Largely copied from our credit purchasing flow [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-webapp/src/packages/cloud/views/credits/CreditsPage/components/RemainingCredits.tsx#L104).

After successfully entering a card, users will be redirected back to the URL they came from with an extra query param: `?stripeCheckoutSuccess=true`

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/EnrollmentModal.tsx
Comment (josephkmh) [inline airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/EnrollmentModal.tsx:55]: This depends on https://github.com/airbytehq/airbyte/issues/21228

File: airbyte-webapp/src/packages/cloud/lib/domain/stripe/types.ts
Comment (josephkmh) [inline airbyte-webapp/src/packages/cloud/lib/domain/stripe/types.ts:6]: This will be a required param for `createCheckout`

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/MailSvg.tsx
Comment (tealjulia) [inline airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/MailSvg.tsx]: I think we lean towards using the actual .svg file rather than exporting it in a .tsx in the codebase

(can use `import { ReactComponent as WhateverYouWantToCallIt } from ....` to use it as a component)

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentButton.tsx
Comment (tealjulia) [inline airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentButton.tsx:26]: We'll need to call this from a span-acting-as-a-button in the inline banner ([here](https://github.com/airbytehq/airbyte/pull/21315)).  Can we split out this logic so it's reusable?

**Review by teallarson — COMMENTED** [review]
- have not tried it locally yet
- The only comment that I feel pretty strongly about is the one in the button component... we may want to utilize the `ModalService` like our other modals do, though you said you have something with a hook that may work.

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentButton.tsx
Comment (ambirdsall) [inline airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentButton.tsx:38]: nit: I believe `onClick={showEnrollmentModal}` is equivalent but slightly (subjectively) nicer and trivially more efficient

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentModal.module.scss
Comment (ambirdsall) [inline airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentModal.module.scss:24]:
```suggestion
  width: variables.$width-modal-sm;
```

**Review by ambirdsall — COMMENTED** [review]
Two fairly trivial edits I'd like to see before merging, but overall looks good.

**Review by ambirdsall — APPROVED** [review]
Approving now: this PR is pretty safe (it adds no user-facing changes, no apparent correctness issues, and all requested edits are small), and so I'd rather not hold up Joey's next work morning waiting for North American time zones to catch up.

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/MailSvg.tsx
Comment (josephkmh) [reply airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/MailSvg.tsx]: For sure, that makes sense. I'll keep it consistent with the rest of our SVGs!

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentButton.tsx
Comment (josephkmh) [reply airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentButton.tsx:38]: True, in this case equivalent and can be simplified - will do. I think the extra anonymous function was leftover when I was experimenting with passing through props to `openModal`.

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentButton.tsx
Comment (josephkmh) [reply airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentButton.tsx]: Good point, I've refactored this into the `useShowEnrollmentModal()` hook which can then be used anywhere. I got rid of the `EnrollmentButton` component, since it became pretty redundant!

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/MailSvg.tsx
Comment (josephkmh) [reply airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/MailSvg.tsx]: https://github.com/airbytehq/airbyte/pull/21396/commits/4e4778fe89823d43525fe164786903bbc02e3f53

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentButton.tsx
Comment (josephkmh) [reply airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/EnrollmentButton/EnrollmentButton.tsx]: Actually ended up removing this component in favor of the `useShowEnrollmentModal()` hook, but I removed some other unnecessary anonymous functions elsewhere!

File: airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/MailSvg.tsx
Comment (tealjulia) [reply airbyte-webapp/src/packages/cloud/components/experiments/FreeConnectorProgram/FreeConnectorProgramEnrollmentButton/MailSvg.tsx]: I'm not sure if there's something weird going on with the diff, but it looks like there may be a .tsx and a .svg for each image now?

**Review by teallarson — APPROVED** [review]
It looks like the .svg images still have the .tsx counterparts as well.  Other than that, looks good to me!

**Pull request merged**
